### PR TITLE
[Enhancement](merge-on-write) add correctness check for the calculation of delete bitmap

### PR DESCRIPTION
**Pull request by bobhan1**
## Proposed changes

Currently, for merge-on-write unique table, the delete bitmap of a rowset will be calculated during flush phase, commit phase and publish phase. In this PR, we add a special mark in every rowset considered when we calculate delete bitmap in these three phases. Before we finally merge the delete bitmap to the table meta's delete bitmap, we will check if all the rowsets contain the special mark to check if we have considered all the rowsets during the above three phases. 
Because the executor can not fail in publish phase if the coordinator have received successful commits info from all the executors, we just print logs if this correctness check failed rather than report a failure.
## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by bobhan1** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 48.24 seconds
 stream load tsv:          505 seconds loaded 74807831229 Bytes, about 141 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          30 seconds loaded 861443392 Bytes, about 27 MB/s
 insert into select:          29.2 seconds inserted 10000000 Rows, about 342K ops/s
 storage size: 17164434966 Bytes

**Comment by bobhan1** [comment]
run buildall

**Comment by bobhan1** [comment]
run buildall

**Comment by bobhan1** [comment]
run buildall

**Comment by bobhan1** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 44.73 seconds
 stream load tsv:          501 seconds loaded 74807831229 Bytes, about 142 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          64 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.3 seconds inserted 10000000 Rows, about 341K ops/s
 storage size: 17167839689 Bytes

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 44.78 seconds
 stream load tsv:          504 seconds loaded 74807831229 Bytes, about 141 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          64 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          30 seconds loaded 861443392 Bytes, about 27 MB/s
 insert into select:          29.0 seconds inserted 10000000 Rows, about 344K ops/s
 storage size: 17163042439 Bytes

File: be/src/olap/tablet_meta.h [hunk 1]
```diff
@@ -302,6 +303,10 @@ class DeleteBitmap {
     using Version = uint64_t;
     using BitmapKey = std::tuple<RowsetId, SegmentId, Version>;
     std::map<BitmapKey, roaring::Roaring> delete_bitmap; // Ordered map
+    constexpr static inline uint32_t INVALID_SEGMENT_ID = std::numeric_limits<uint32_t>::max();
```
Comment (zhannngchen) [inline be/src/olap/tablet_meta.h:306]: use max()-1

File: be/src/olap/tablet_meta.h [hunk 2]
```diff
@@ -302,6 +303,10 @@ class DeleteBitmap {
...
+    constexpr static inline uint64_t INVALID_VERSION = std::numeric_limits<uint64_t>::max();
```
Comment (zhannngchen) [inline be/src/olap/tablet_meta.h:307]: use 0

**Comment by bobhan1** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 45.77 seconds
 stream load tsv:          505 seconds loaded 74807831229 Bytes, about 141 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          64 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.4 seconds inserted 10000000 Rows, about 340K ops/s
 storage size: 17169261124 Bytes

**Comment by bobhan1** [comment]
run clickbench

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 47.87 seconds
 stream load tsv:          549 seconds loaded 74807831229 Bytes, about 129 MB/s
 stream load json:         21 seconds loaded 2358488459 Bytes, about 107 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.0 seconds inserted 10000000 Rows, about 344K ops/s
 storage size: 17162250585 Bytes

File: be/src/olap/tablet.h [hunk 3]
```diff
@@ -545,6 +546,12 @@ class Tablet : public BaseTablet {
 
     void set_binlog_config(BinlogConfig binlog_config);
 
+    void add_sentinel_mark_to_delete_bitmap(DeleteBitmapPtr delete_bitmap,
```
Comment (zhannngchen) [inline be/src/olap/tablet.h:549]: These methods should be private

File: be/src/olap/tablet.cpp [hunk 4]
```diff
@@ -3637,4 +3657,46 @@ Status Tablet::calc_delete_bitmap_between_segments(
...
+
+void Tablet::remove_sentinel_mark_from_delete_bitmap(DeleteBitmapPtr delete_bitmap) {
+    for (auto& [key, bitmap] : delete_bitmap->delete_bitmap) {
+        bitmap.remove(DeleteBitmap::ROWSET_SENTINEL_MARK);
+    }
+}
+
+Status Tablet::check_delete_bitmap_correctness(DeleteBitmapPtr delete_bitmap, int64_t max_version) const {
+    std::map<RowsetId, bool> result;
+    RowsetIdUnorderedSet rowsets = all_rs_id(max_version);
```
Comment (zhannngchen) [inline be/src/olap/tablet.cpp:3676]: Should hold the _meta_lock

File: be/src/olap/tablet.cpp [hunk 5]
```diff
@@ -3318,6 +3332,12 @@ Status Tablet::update_delete_bitmap(const RowsetSharedPtr& rowset,
               << ", cur max_version: " << cur_version << ", transaction_id: " << txn_id
               << ", cost: " << watch.get_elapse_time_us() << "(us), total rows: " << total_rows;
 
+    if (do_correctness_check) {
+        // check if all the rowset has ROWSET_SENTINEL_MARK
+        RETURN_IF_ERROR(check_delete_bitmap_correctness(delete_bitmap, cur_version - 1));
```
Comment (zhannngchen) [inline be/src/olap/tablet.cpp:3337]: consider reuse the variable `cur_rowset_ids`?

File: be/src/olap/tablet.cpp (hunk 4 above)
Comment (zhannngchen) [reply be/src/olap/tablet.cpp:3676]: if you can reuse the `rowset_ids` from caller, then you don't need to acquire the lock here again

File: be/src/olap/tablet.cpp [hunk 6]
```diff
@@ -3637,4 +3657,46 @@ Status Tablet::calc_delete_bitmap_between_segments(
...
+Status Tablet::check_delete_bitmap_correctness(DeleteBitmapPtr delete_bitmap, int64_t max_version) const {
+    std::map<RowsetId, bool> result;
+    RowsetIdUnorderedSet rowsets = all_rs_id(max_version);
+    for (const auto& rowsetid : rowsets) {
+        result.emplace(rowsetid, false);
+    }
+    for (const auto& [key, bitmap] : delete_bitmap->delete_bitmap) {
+        auto it = result.find(std::get<0>(key));
+        if (it == result.end()) {
+            LOG(WARNING) << "can't find rowsetid when checking delete bitmap correctness";
```
Comment (zhannngchen) [inline be/src/olap/tablet.cpp:3683]: print the missing rowset_id

**Comment by bobhan1** [comment]
run buildall

**Review by zhannngchen — DISMISSED** [review]
LGTM

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 47.52 seconds
 stream load tsv:          512 seconds loaded 74807831229 Bytes, about 139 MB/s
 stream load json:         21 seconds loaded 2358488459 Bytes, about 107 MB/s
 stream load orc:          64 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          30.1 seconds inserted 10000000 Rows, about 332K ops/s
 storage size: 17162153014 Bytes

**Comment by bobhan1** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 46.18 seconds
 stream load tsv:          515 seconds loaded 74807831229 Bytes, about 138 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.1 seconds inserted 10000000 Rows, about 343K ops/s
 storage size: 17162455773 Bytes

**Comment by bobhan1** [comment]
run buildall

File: be/src/olap/tablet.cpp [hunk 7]
```diff
@@ -3318,6 +3333,13 @@ Status Tablet::update_delete_bitmap(const RowsetSharedPtr& rowset,
               << ", cur max_version: " << cur_version << ", transaction_id: " << txn_id
               << ", cost: " << watch.get_elapse_time_us() << "(us), total rows: " << total_rows;
 
+    if (do_correctness_check) {
```
Comment (liaoxin01) [inline be/src/olap/tablet.cpp:3336]: we can just use `rowset->num_rows() > 0` here, don't need to pass do_correctness_check param.

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 47.23 seconds
 stream load tsv:          515 seconds loaded 74807831229 Bytes, about 138 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          64 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.3 seconds inserted 10000000 Rows, about 341K ops/s
 storage size: 17162126005 Bytes

**Comment by bobhan1** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 45.57 seconds
 stream load tsv:          511 seconds loaded 74807831229 Bytes, about 139 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          64 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.0 seconds inserted 10000000 Rows, about 344K ops/s
 storage size: 17162027403 Bytes

**Comment by bobhan1** [comment]
run p0

File: be/src/olap/tablet.cpp [hunk 8]
```diff
@@ -2977,6 +2977,17 @@ Status Tablet::calc_segment_delete_bitmap(RowsetSharedPtr rowset,
...
     DCHECK_EQ(total, row_id) << "segment total rows: " << total << " row_id:" << row_id;
 
+    RowsetIdUnorderedSet rowsetids;
+    for (const auto& rowset : specified_rowsets) {
+        rowsetids.emplace(rowset->rowset_id());
+        LOG(INFO) << "[tabletID:" << tablet_id() << "]"
+                  << "[_add_sentinel_mark_to_delete_bitmap][end_version:" << end_version << "]"
+                  << "add:" << rowset->rowset_id();
+    }
+    if (config::enable_merge_on_write_correctness_check) {
```
Comment (liaoxin01) [inline be/src/olap/tablet.cpp:2987]: Place the code from line 2980 to line 2986 within if condition.

**Comment by bobhan1** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 46.78 seconds
 stream load tsv:          512 seconds loaded 74807831229 Bytes, about 139 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          70 seconds loaded 1101869774 Bytes, about 15 MB/s
 stream load parquet:          30 seconds loaded 861443392 Bytes, about 27 MB/s
 insert into select:          29.0 seconds inserted 10000000 Rows, about 344K ops/s
 storage size: 17162138571 Bytes

**Review by liaoxin01 — APPROVED** [review]
LGTM

**Comment by bobhan1** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 44.99 seconds
 stream load tsv:          513 seconds loaded 74807831229 Bytes, about 139 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          30 seconds loaded 861443392 Bytes, about 27 MB/s
 insert into select:          29.2 seconds inserted 10000000 Rows, about 342K ops/s
 storage size: 17162282284 Bytes

**Review by zhannngchen — DISMISSED** [review]
LGTM

**Comment by zhannngchen** [comment]
run buildall

**Comment by bobhan1** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 46.46 seconds
 stream load tsv:          513 seconds loaded 74807831229 Bytes, about 139 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          30 seconds loaded 861443392 Bytes, about 27 MB/s
 insert into select:          29.2 seconds inserted 10000000 Rows, about 342K ops/s
 storage size: 17161955253 Bytes

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 46.65 seconds
 stream load tsv:          512 seconds loaded 74807831229 Bytes, about 139 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          32 seconds loaded 861443392 Bytes, about 25 MB/s
 insert into select:          29.0 seconds inserted 10000000 Rows, about 344K ops/s
 storage size: 17162510251 Bytes

**Comment by bobhan1** [comment]
run arm

**Review by zhannngchen — APPROVED** [review]
LGTM

**Pull request merged**